### ISSUE TITLE
Reduce use of ConstraintLayout

### ISSUE THREAD

**Issue by fynngodau** [corona-warn-app/cwa-app-android#1855]
When investigating why the privacy policy screen is taking so long to load, I've learned that `ConstraintLayout`s appear to be quite inefficient (in the mentioned case, replacing it with a `LinearLayout` has helped significantly, see #1854).

There is oftentimes a better alternative to `ConstraintLayout` which also makes the layout easier to read and maintain (see, for example, #1740). Therefore, I would like to suggest avoiding the use of `ConstraintLayout` and to prefer using simpler layouts like `LinearLayout` or `RelativeLayout` unless there is a good reason not to.

Should I come across other locations where there is a particular benefit to avoiding `ConstraintLayout`, I will of course open further PRs.

In specific, I suspect that startup time could be improved by a more simple home layout.

**Comment by d4rken** [comment]
> In specific, I suspect that startup time could be improved by a more simple home layout.

I agree. We plan to flatten the home layout and remove all uses of "visible"/"gone" on included layouts by using a `RecyclerView` to display the elements.

**Comment by fynngodau** [comment]
That's a great idea, `RecyclerView` is very well fit for this purpose.

For dealing with `RecyclerView`s with different items that must be rendered as different views, I wrote [a utility library](https://codeberg.org/fynngodau/TypedRecyclerView) that allows for a more object-oriented approach over passing the `RecyclerView.Adapter` superclass different `int`s for each view (which usually results in one single rather ugly class that handles all the different views). I find it useful for cases like these in my own projects, but maybe it's not well fit for this project, I'm not sure.

**Comment by d4rken** [comment]
I've written a `ModularAdapter` pattern a few times already which I will likely re-use here as I know it plays nicely with `DiffUtil`, which we will need to update cards independently.

..hm don't have it in an open repo to show, but the basic idea is to allow adding modules to the adapter, where a module can be  something that binds data to a viewholder, or creates a viewholder or maps clicks/longclicks to viewholders.

e.g.
```kotlin
        modules.add(TypedVHCreator(0, { data[it].config is AppRestoreConfig }) { AppConfigVH(it) })
        modules.add(TypedVHCreator(1, { data[it].config is FilesRestoreConfig }) { FileConfigVH(it) })
```

I'm not stoked about the "ID" assignment for VHs in my solution though.
I liked the pattern (in your utility library) that you used in the `positionInSet`  helper method :heart: , that might be something I could apply. I'll ping you on the PR that introduces the recycler for the home screen.

**Comment by fynngodau** [comment]
I'm looking forward to having a look at your solution!

**Comment by fynngodau** [comment]
I'm closing this issue because a more simple home layout like I suggested is now implemented with #1912, and because CWA [uses a newer version of `ConstraintLayout`](https://github.com/corona-warn-app/cwa-app-android/pull/2063/files#diff-264ff2dbc2ac999887d86a9dc8b73fb2380a6e76aeeda85b480bc7a5dc1470b9R316) which potentially has better performance (I didn't specifically try that though); causing this issue to have become too vague.
